cirq/circuits/circuit.py

```diff
@@ -407,6 +408,208 @@ def _prev_moment_available(
+            enqueue_next(start_qubit, start_moment)
+
+        while queue:
+            cur_moment, cur_op = queue.dequeue()
+            for q in cur_op.qubits:
+                if (q in start_frontier and
+                        cur_moment >= start_frontier[q] and
+                        q not in end_frontier):
+                    active.add(q)
+
+            continue_past = (
+                    cur_op is not None and
+                    active.issuperset(cur_op.qubits) and
+                    not is_blocker(cur_op)
+            )
+            if continue_past:
+                for q in cur_op.qubits:
+                    enqueue_next(q, cur_moment + 1)
+            else:
+                for q in cur_op.qubits:
+                    if q in active:
+                        end_frontier[q] = cur_moment
+                        active.remove(q)
+
+        return end_frontier
+
+    def findall_operations_between(self,
+                                   start_frontier: Dict[ops.QubitId, int],
+                                   end_frontier: Dict[ops.QubitId, int],
+                                   omit_crossing_operations: bool = False
+                                   ) -> List[Tuple[int, ops.Operation]]:
+        """Finds operations between the two given frontiers.
+
+        If a qubit is in `start_frontier` but not `end_frontier`, its end index
+        defaults to the end of the circuit. If a qubit is in `end_frontier` but
+        not `start_frontier`, its start index defaults to the start of the
+        circuit. Operations on qubits not mentioned in either frontier are not
+        included in the results.
+
+        Args:
+            start_frontier: Just before where to start searching for operations,
+                for each qubit of interest. Start frontier indices are
+                inclusive.
+            end_frontier: Just before where to stop searching for operations,
+                for each qubit of interest. End frontier indices are exclusive.
+            omit_crossing_operations: Determines whether or not operations that
+                cross from a location between the two frontiers to a location
+                outside the two frontiers are included or excluded. (Operations
+                completely inside are always included, and operations completely
+                outside are always excluded.)
+
+        Returns:
+            A list of tuples. Each tuple describes an operation found between
+            the two frontiers. The first item of each tuple is the index of the
+            moment containing the operation, and the second item is the
+            operation itself. The list is sorted so that the moment index
+            increases monotonically.
+        """
+        result = BucketPriorityQueue[ops.Operation](drop_duplicates=True)
+
+        involved_qubits = set(start_frontier.keys()) | set(end_frontier.keys())
+        for q in ops.QubitOrder.DEFAULT.order_for(involved_qubits):
```

Comment:
should this method take in a qubit order that can be used here?

Reply:
No, the sorting is just to ensure a deterministic output. The ordering doesn't matter, as long as it's consistent for testing. Added a comment saying such.

Also, I switched from `ops.QubitOrder.DEFAULT.order_for` to just `sorted`, now that that finally works. :smile: